packages/block-editor/src/store/selectors.js

```diff
@@ -1537,3 +1539,128 @@ export function didAutomaticChange( state ) {
 export function __unstableGetSelectedMountedBlock( state ) {
 	return state.selectedMountedBlock;
 }
+
+/**
+ * Get the label for the block, usually this is either the block title,
+ * or the value of the attribute denoted by the block's `label` property.
+ *
+ * @param {Object} state    Store state.
+ * @param {string} clientId ClientId for the block.
+ *
+ * @return {string} The block label.
+ */
+export function getBlockLabel( state, clientId ) {
+	const blockName = getBlockName( state, clientId );
+
+	const {
+		__experimentalLabel: labelAttribute,
+		title,
+	} = getBlockType( blockName );
+
+	if ( ! labelAttribute ) {
+		return title;
+	}
+
+	const attributes = getBlockAttributes( state, clientId );
+	const label = attributes[ labelAttribute ];
+
+	if ( ! label ) {
+		return title;
+	}
+
+	// Strip any HTML (i.e. RichText formatting) before returning.
+	return toPlainText( label );
+}
+
+/**
+ * Get a label for the block for use by screenreaders, this can include the block title, the
+ * position of the block, and the value of the `getAccessibilityLabel` function if it's specified.
+ *
+ * @param {Object} state    Store state.
+ * @param {string} clientId ClientId for the block.
+ * @param {"horizontal"|"vertical"} [direction="vertical"] The direction blocks are layed out in,
+ *                                                         can be 'horizontal' or 'vertical'.
+ *
+ * @return {string} The accessibility label for the block.
+ */
+export function getAccessibleBlockLabel( state, clientId, direction = 'vertical' ) {
+	const blockName = getBlockName( state, clientId );
+
+	const {
+		__experimentalGetAccessibilityLabel: getAccessibilityLabel,
```

Comment:
I think we need some more consistency with how these things are named.

We reference both "accessible" and "accessibility" labels? Can we pick one?

Or also: Why do we need to qualify it? Can't we just call it a label? I worry the qualifier might make this some overly specialized thing, if all we truly care about is having some meaningful way to describe the instance of the block, which can be (but not necessarily _only be_) used as its accessible label. This ties to what I mention in https://github.com/WordPress/gutenberg/pull/18132/files#r365328275 and https://github.com/WordPress/gutenberg/pull/18132/files#r367984986.

Reply:
> We reference both "accessible" and "accessibility" labels? Can we pick one?

True, I can sort out the consistency.

> Or also: Why do we need to qualify it? Can't we just call it a label? I worry the qualifier might make this some overly specialized thing, if all we truly care about is having some meaningful way to describe the instance of the block, which can be (but not necessarily only be) used as its accessible label.

I'm not sure I fully understand this criticism. It's the same reason we often use `aria-label`, `aria-description` or `screen-reader-text` to add extra information over and above the existing visual labels. Sometimes that's because labels aren't present, but there are plenty of cases where they are, but are not adequate for screen reader usage (block movers are an example).

Replying to your linked comment:

> Even in the implementations we've revised so far, we're only ever handling the "accessibility" case. So what's the point of "visual"? 

So there's a bit of history here. Originally, the mobile devs implemented an `__experimentalGetAccessibilityLabel ` for blocks. That's why the majority of the existing cases are accessibility labels.

The navigation block is the only block that implements both a 'visual' label at the moment. There was a specific requirement to do that at the time.

Now I'm attempting to tidy things up (but perhaps making things worse!). My original PR #18132 was an attempt to try to consolidate these APIs into a single label, but the reality is that the accessible label often doesn't work visually, and vice versa.

Paragraph and Heading are two instances where there's probably a bit more information in the accessibility label than we'd consider showing visually. But they've already been assessed as being useful for screen readers.

Perhaps the mistake is that they shouldn't both be called 'labels' as that implies an association. Maybe one should be `getAccessibilityDescription` and the other a label? They are used differently at the moment (one in navigation mode, the other in the block navigator) so this probably makes sense.

Reply:
> Perhaps the mistake is that they shouldn't both be called 'labels' as that implies an association. Maybe one should be getAccessibilityDescription and the other a label? They are used differently at the moment (one in navigation mode, the other in the block navigator) so this probably makes sense.

The trouble with AccessibilityDescription is that this is a separate property on Accessibility objects on various platforms. Windows, Linux, and Mac have an AccessibleDescription which is *secondary* information to the AccessibilityLabel. AccessibilityLabel is the primary thing that says what a button or other thing is named, or will do when executed. The description is additional information. On iOS, this is called AccessibilityHint. It is what VoiceOver speaks after a delay, on a Home Screen element, for example, this is "Double tap to open". In mobile Gutenberg, this is often used to communicate helpful things like "Double-tap to add a block" or instructions pertaining to a particular block type.

So in terms of Accessibility terminology, AccessibilityLabel is the perfectly correct and unambiguous term to use here.

Reply:
Thanks for the background @MarcoZehe.

Reply:
> I'm not sure I fully understand this criticism. It's the same reason we often use `aria-label`, `aria-description` or `screen-reader-text` to add extra information over and above the existing visual labels.

I think the parallels you mention are appropriate, but I see them as backing up my worries. I may be mistaken, but the way I understand those attributes is as something of an escape hatch where the design of a visual label would not sufficient on its own in providing a meaningful name. And that, whenever possible, we should optimize for a label which can be well-understood in all contexts (visual or non-visual). 

It may very well be true that in some cases we'd want to present these labels with some added contextual clarification (in other words, maybe we do want to use them in `aria-label`, perhaps with some prefix or suffix text). But are those additional details going to be distinct per-block, or something generic that we can decide at a framework level? In the latter case, it affords us the opportunity to remove this from consideration of the block implementer.

At the end of the day, what conceptual details are we wanting the block to provide here? Is it just a short description of the contents of that block? My initial reaction here is that this is something which can be implemented without tying it so specifically to the implementation detail of how that label is used. Or, at least, based on my understanding of how we're currently proposing to use that label, it seemed perhaps premature. 

>but the reality is that the accessible label often doesn't work visually, and vice versa.

I admit that I'm not fully aware of all of the history here, and that this is fair reasoning for why they be treated distinctly. But I think my approaching this from a perspective of ignorance might be useful in considering to reevaluate this. We should always be open to the possibility that we've been doing it wrong all along. What is it specifically that makes one appropriate visually and not for screen readers, and vice-versa, like you mention? Are those differences reasonable, or symptomatic of a larger problem? 

If it helps, my motivation here is largely to avoid forcing the block implementer needing to be aware of a distinction or needing to make a choice (especially where a choice could have negative accessibility implications if done incorrectly), and to distill the requirement to one which maximizes reusability.